src/TableFunctions/TableFunctionLoop.cpp

```diff
@@ -0,0 +1,144 @@
+
+    if (args.size() == 1)
+    {
+        if (const auto * id = args[0]->as<ASTIdentifier>())
+        {
+            String id_name = id->name();
+
+            size_t dot_pos = id_name.find('.');
+            if (dot_pos != String::npos)
+            {
+                database_name_ = id_name.substr(0, dot_pos);
+                table_name_ = id_name.substr(dot_pos + 1);
+            }
+            else
+            {
+                table_name_ = id_name;
+            }
+        }
+        else if (const auto * func = args[0]->as<ASTFunction>())
+        {
+            inner_table_function_ast = args[0];
+        }
+        else
+        {
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Expected identifier or function for argument 1 of function 'loop', got {}", args[0]->getID());
+        }
+    }
+    // loop(database, table)
+    else if (args.size() == 2)
+    {
+        args[0] = evaluateConstantExpressionForDatabaseName(args[0], context);
+        args[1] = evaluateConstantExpressionOrIdentifierAsLiteral(args[1], context);
+
+        database_name_ = checkAndGetLiteralArgument<String>(args[0], "database");
+        table_name_ = checkAndGetLiteralArgument<String>(args[1], "table");
+    }
+    else
+    {
+        throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Table function 'loop' must have 1 or 2 arguments.");
+    }
+}
+
+ColumnsDescription TableFunctionLoop::getActualTableStructure(ContextPtr /*context*/, bool /*is_insert_query*/) const
+{
+    return ColumnsDescription();
+}
+
+StoragePtr TableFunctionLoop::executeImpl(
+    const ASTPtr & /*ast_function*/,
+    ContextPtr context,
+    const std::string & table_name,
+    ColumnsDescription cached_columns,
+    bool is_insert_query) const
+{
+    StoragePtr storage;
+    if (!table_name_.empty())
+    {
+        String database_name = database_name_;
+        if (database_name.empty())
+            database_name = context->getCurrentDatabase();
+
+        auto database = DatabaseCatalog::instance().getDatabase(database_name);
+        storage = database->tryGetTable(table_name_, context);
+        if (!storage)
+            throw Exception(ErrorCodes::UNKNOWN_TABLE, "Table '{}' not found in database '{}'", table_name_, database_name);
+    }
+
+    else
+    {
+        auto inner_table_function = TableFunctionFactory::instance().get(inner_table_function_ast, context);
+        storage = inner_table_function->execute(
+            inner_table_function_ast,
+            context,
+            table_name,
+            std::move(cached_columns),
+            is_insert_query);
+    }
+    auto res = std::make_shared<StorageLoop>(
+        StorageID(getDatabaseName(), table_name),
+        storage
+        );
+    res->startup();
+    return res;
+}
+
+void registerTableFunctionLoop(TableFunctionFactory & factory)
+{
+    factory.registerFunction<TableFunctionLoop>();
```

Comment:
Could you please add description here as it is done here: https://github.com/ClickHouse/ClickHouse/blob/03b47ddcbdf84f5af39775f52051352e17fe26f5/src/TableFunctions/TableFunctionAzureBlobStorage.cpp#L388-L389